### PR TITLE
Return structs from function calls

### LINKED ISSUE #283: Return structs in read functions
With the experimental ABIv2 we can return Structs out of function calls (as well as pass those in). It it worth exploring as a design in our contracts potentially with the following benefits:

- Enforces return data structures to comply with those declared in storage. This was the reason behind the misalignment highlighted in `getSkill` return value issue here  https://github.com/JoinColony/colonyNetwork/issues/249
- Simplifies upgrades as extending a struct (e.g. adding a new member) won't necessitate upgrade of the function return parameter type - subject to testing

Candidates for this refactoring are essentially all getter functions where a struct is returned: `getSkill`, `getTask`, `getDomain`, `getTaskWorkRatings`, `getTaskRole`, `getRewardPayoutInfo`, `getUserLock` etc

### PR DESCRIPTION
**Pull request by elenadimitrova**
Closes #283

Depends on the truffle upgrade in #435 which introduces support for [`ABIEncoderV2`](https://solidity.readthedocs.io/en/v0.5.0/layout-of-source-files.html?highlight=ABIEncoderV2#abiencoderv2) including returning structs from functions. 

Here we demonstrate how `getDomain` can return the `Domain` struct. In tests this is backward compatible as it currently returns the two sets for the same values, once as plain data types and secondly as named parameters. For instance `getDomain` returns: 
```
[ '4', '1', skillId: '4', potId: '1' ]
```
so we can do both `domain[0]` and `domain.skillId`.

Overall to proceed we should consider whether using `ABIEncoderV2` is production ready for us. There are varying opinions of this out there, [`ethers` for example](https://blog.ricmoo.com/solidity-abiv2-a-foray-into-the-experimental-a6afd3d47185) is (as always) on the cautious side, while `0x` are sufficiently confident to [roll it out in production back in February](https://blog.0xproject.com/development-update-6-february-2018-8d6f6018eabf).

At present remix IDE is not yet working for the new 0.5.0 compiler so cannot try this with the new red hot `solc` release. However there's nothing in the release notes mentioning support for this and I'm more certain than not that returning structs is still only in the experimental `ABIEncoderV2`.

### CODE REVIEW

**Comment by area** [comment]
Given we're only proposing we use it for reading from a contract for now, rather than passing things in (though I can see it being useful for packing the inputs to `respondToChallenge` down even further), I can't see an issue with it still being considered experimental?

File: contracts/ColonyFunding.sol [hunk 1]
```diff
@@ -15,8 +15,9 @@
   along with The Colony Network. If not, see <http://www.gnu.org/licenses/>.
 */
 
-pragma solidity ^0.4.23;
+pragma solidity ^0.4.22;
```
Comment (elenadimitrova) [inline contracts/ColonyFunding.sol]: ? 🤔

Comment (area) [reply contracts/ColonyFunding.sol]: No clue what happened here :confused:

File: contracts/ColonyNetwork.sol [hunk 2]
```diff
@@ -70,9 +71,16 @@ contract ColonyNetwork is ColonyNetworkStorage {
...
+  function getSkill(uint256 _skillId) public view returns (Skill skill) {
+    skill = skills[_skillId];
+  }
+
+  function isGlobalSkill(uint256 _skillId) public view returns (bool globalSkill) {
+    globalSkill = skills[_skillId].globalSkill;
+  }
+
+  function getSkillNParents(uint256 _skillId) public view returns (uint256 nParents) {
+    nParents = skills[_skillId].nParents;
```
Comment (elenadimitrova) [inline contracts/ColonyNetwork.sol]: I can't understand why this is needed?

**Pull request merged**